### PR TITLE
feat(binder): add concat

### PR DESCRIPTION
**Pull request by kwannoel**
## What's changed and what's your intention?

As per title.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

related: https://github.com/singularity-data/risingwave/issues/3006

### CODE REVIEW

**Comment by neverchanje** [comment]
pls add some e2e tests :)

**Comment by kwannoel** [comment]
> pls add some e2e tests :)

Currently `concat` is not fully supported, because frontend hasn't done binding. Would like to add e2e tests along with frontend changes.

**Comment by kwannoel** [comment]
> > pls add some e2e tests :)
> 
> Currently `concat` is not fully supported, because frontend hasn't done binding. Would like to add e2e tests separately along with frontend changes in a different PR, to keep changes small

I guess I can do both in this PR too, changes should be straightforward to review

File: src/frontend/test_runner/tests/testdata/expr.yaml [hunk 1]
```diff
@@ -256,3 +256,25 @@
...
+    create table t (v1 float);
+    select concat(v1) from t;
+  batch_plan: |
+    BatchExchange { order: [], dist: Single }
+      BatchProject { exprs: [Concat($0::Varchar)] }
+        BatchScan { table: t, columns: [v1] }
+- sql: |
+    create table t (v1 int);
+    select concat() from t;
+  binder_error: 'Bind error: Function `ConcatWs` takes at least 1 arguments (0 given)'
```
Comment (kwannoel) [inline src/frontend/test_runner/tests/testdata/expr.yaml:280]: typo

File: src/frontend/src/expr/function_call.rs [hunk 2]
```diff
@@ -124,6 +124,21 @@ impl FunctionCall {
                 }
                 align_types(inputs.iter_mut())
             }
+            ExprType::Concat => {
+                if inputs.is_empty() {
+                    return Err(ErrorCode::BindError(format!(
+                        "Function `ConcatWs` takes at least {} arguments ({} given)",
```
Comment (kwannoel) [inline src/frontend/src/expr/function_call.rs:130]: typo

File: e2e_test/batch/functions/concat.slt [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+query T
+select concat(NULL);
+----
+
+
+query T
+select concat(NULL, NULL);
+----
+
```
Comment (kwannoel) [inline e2e_test/batch/functions/concat.slt:14]: https://www.sqlite.org/sqllogictest/doc/trunk/about.wiki
> Empty strings are rendered as "(empty)".

seems like `sqllogictest-rs` differs from spec for empty string. Need to verify.

Comment (kwannoel) [reply e2e_test/batch/functions/concat.slt:14]: Yes result from `select concat(NULL, NULL);` is like so:
```
dev=> select concat(NULL) = '';
 ?column? 
----------
 t
(1 row)
```

File: src/expr/src/expr/expr_concat.rs [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+impl<'a> TryFrom<&'a ExprNode> for ConcatExpression {
+    type Error = ExprError;
+
+    fn try_from(prost: &'a ExprNode) -> Result<Self> {
+        ensure!(prost.get_expr_type().unwrap() == Type::Concat);
+
+        let ret_type = DataType::from(prost.get_return_type().unwrap());
+        assert_eq!(ret_type, DataType::Varchar);
```
Comment (kwannoel) [inline src/expr/src/expr/expr_concat.rs:101]: use ensure_eq instead.

**Review by xiangjinwu — COMMENTED** [review]
It works. But is it possible to avoid repeating ourselves between `concat_ws` and `concat`? What about mapping it to `concat_ws` with literal empty string as separator in binder?

**Comment by kwannoel** [comment]
> It works. But is it possible to avoid repeating ourselves between concat_ws and concat? What about mapping it to concat_ws with literal empty string as separator in binder?

Yes that's a good idea.

**Review by xiangjinwu — APPROVED** [review]

**Pull request merged**